### ISSUE TITLE
Improve reproducibility of artifacts

### ISSUE THREAD

**Issue by msarahan** [conda/conda-build#2140]
@bollwyvl raised this privately.  He would like conda packages to be completely verifiable.  There's some good guidelines on this at https://reproducible-builds.org/

@bollwyvl has some preliminary work implemented in a jupyter notebook that I'm happy to share if anyone would like to see.

**Comment by bollwyvl** [comment]
Thanks, @msarahan! Also thanks for not just posting that notebook... it's of course broken 😈 .

In summary:

My early work has a lot of shortcomings:
- only tried with `conda-build` 2.1.6
- only tried with pure python packages: a toy example, and `conda-build` 2.1.16 itself
- only tried on OSX

Observations in no particular order: 
- [x] `info/index.json` wasn't being sorted (has been fixed on master) 🎆 
- [ ] tarballs preserve timestamp and owner
  - easy enough to fix, either by post-processing or something
- [ ] `info/files` is not [sorted](https://github.com/conda/conda-build/blob/56e8ceba1d016a23ba419ad93a6d77b017c6ca14/conda_build/build.py#L324)
- [ ] `info/recipe/meta.yaml` has non-deterministic ordering of sets (mainly package spec listings)
- [ ] `versioneer.py` (and maybe other things that hack how pip/setuptools do their thing) apparently don't respect `PYTHONHASHSEED` and `DETERMINISTIC_BUILD`, so they end up with 1-2 bytes of header differences in `pyc` files
- [ ]  `info/about.json`
  - [ ] there's no easy way to extract`root_pkgs` directly to reproduce the build environment
  - [ ] `env_vars` includes username and other transient path information

In working on this, I have found [`diffoscope`](https://diffoscope.org/) super valuable, and am trying to get it built for conda, but don't have any success there yet!

**Comment by mingwandroid** [comment]
Good to see someone paying attention to this stuff!

**Comment by bollwyvl** [comment]
An update: diffoscope PR on conda-forge:
- https://github.com/conda-forge/staged-recipes/pull/3282
  - currently blocked by `libarchive-c` https://github.com/conda-forge/staged-recipes/pull/3281

**Comment by bollwyvl** [comment]
@mandeep you are a reproducibility-enhancing MACHINE. 🎸 on.

Been a bit heads-down, but just updated conda-forge/staged-recipes#3281 and conda-forge/staged-recipes#3282 with some naming issues, extra deps, and typos on my part. Also started sneaking in:
```bash
export DETERMINISTIC_BUILD=1
export PYTHONHASHSEED=0
```
...just to see what happens. Of course, I'd love to remove those things if `conda-build` could start setting them as a matter of course in the temporary build-time scripts!

Additionally, I've started up a reprotest recipe: https://github.com/conda-forge/staged-recipes/pull/3358

**Comment by bollwyvl** [comment]
Ha, again reading too much internet will make me dumb. 

> 
  ```bash
  export DETERMINISTIC_BUILD=1
  ```

Apparently, this a nix-specific thing, which they use (with a patch to deep bits of cpython) to clobber the timestamp `pyc` header bits. Presumably #2234 will fulfill that need... though it may be necessary to reset timestamps _after_ running any recipe-provided patches and _before_ running the build scripts.

**Comment by mandeep** [comment]
@bollwyvl I've been reading a lot into this as well and it seems there's a bit more to do on our side regarding timestamps. I checked the pyc files to see if the magic number was being changed and everything looked okay there, which makes me believe that the difference in bytecode is due to timestamps. #2234 won't fix this unfortunately. I think we need to introduce the $DETERMINISTIC_BUILD and $PYTHONHASHSEED environment variables. The PR at python/cpython#296 has some good discussion on this.

**Comment by bollwyvl** [comment]
@mandeep the rabbit hole is deep indeed! Unfortunately I can't contribute a whole lot right now, but I definitely want to revisit my reproducibly-building-conda-build once some more of your stuff lands!

**Comment by bollwyvl** [comment]
Did some more work on the proposed `conda-build --reproduce`. This is indeed sticky. The state of what you end up with is dependent on many steps: `.condarc`s, command line args, etc.

Those seem reasonable; however, the recipe is a bit tougher. 

If you try to reuse the `info/recipe/meta.yaml.template` (by copying or something), it re-templates and runs the solver again, so you'll (maybe) get newer versions. This also regenerates `.template` and (potentially) injects (local) [timestamps](https://github.com/conda/conda-build/blob/master/conda_build/build.py#L223) into the output `meta.yaml`! 

If you use the `info/recipe/meta.yaml`, you definitely get a different output tarball, since the original `meta.yaml`-for-humans will have been lost in the churn.

Starting from (and including) the "human-centric" (multi-)recipe (and `build_config`), it is reasonable to expect some changes between builds, and as such to get different "final" `meta.yaml`s out. However, to reproduce _the exact build env_ and still get out the same "final" `meta.yaml`, conda-build could:
- special-case around recipe dirs that contain a `meta.yaml` and a `meta.yaml.template`
- or offer something like `--no-template`

**Comment by bollwyvl** [comment]
So some more thinking on this: 

- Some crucial things, like where packages are sourced from when `conda-build` is run, aren't captured anywhere.
  - once a package is reproducible, and built from reproducible packages, this won't matter anway, as the crypto hash will be the same... but we have to get to _that_ point first...
- the relationship between the source `meta.yaml` and the `meta.yaml.template` and `meta.yaml` in `info/recipe/` is tricky for reproducing from a tarball...
- probably some more stuff currently hidden by the above...

So in the near term, it might make sense to start an experimental wrapper, e.g. `conda-build-reproducible` that uses the `conda-build` API, and yields packages _compatible_ with the rest of the `conda` ecosystem, but does pre-and-post processing to move some information around in order to make more guarantees. i.e. `buildinfo`, as discussed on #2239.

Continuing down the pipedream:

The "pop quiz" would be that such a tool could, for at least `linux-64` and `python3.6`, reproducibly rebuild the 43 packages in the `conda-build` runtime dependency chain, the wrapper itself, as well as the _TBD_ packages in the build dependencies of those, _ad-hopefully-not-infinitum_.

Once that was demonstrated, and which I am not sure has even been explored yet, might be a variant of `constructor`, e.g. `reconstructor` that could build reproducible appliances from reproducible packages, and similarly rebuild given a `reconstructor` appliance and a `.buildinfo`.

The "midterm exam" from this would be a `reconstructor` appliance for `linux-64`, self-hosted in the sense that (given sources) it could reproduce all of the ~45 packages in itself, etc. and (finally) itself.

I'm not sure what the "final exam" would be, presumably other architectures!

**Comment by jjhelmus** [comment]
Looks like Python 3.7 may provide a method to create deterministic builds via the accepted [PEP 552 -- Deterministic pycs](https://www.python.org/dev/peps/pep-0552/)

**Comment by bollwyvl** [comment]
> Python 3.7 may provide a method to create deterministic .pyc files

Very exciting!

I've been looking for a chance to pick this back up! Perhaps there's a confluence of py37, [conda-build and conda-forge](https://github.com/conda-forge/conda-forge.github.io/issues/460) and something like [miniforge](https://github.com/jakirkham/miniforge/issues/7) that will be a better "midterm" than a 3.6-based solution could be, given the number of python recipes that would be needed to be updated.

**Comment by leofang** [comment]
xref: #4762
